pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -140,17 +143,113 @@ func readConfig(config io.Reader) (Config, error) {
 	return cfg, err
 }
 
+// parseMetadataUUID reads JSON from OpenStack metadata server and parses
+// instance ID out of it.
+func parseMetadata(jsonData []byte) (string, error) {
+	var jsonif interface{}
+	err := json.Unmarshal(jsonData, &jsonif)
+	if err != nil {
+		return "", err
+	}
+
+	// We should receive an object with { 'uuid': '<uuid> } and couple of other
+	// properties (which we ignore).
+	obj, ok := jsonif.(map[string]interface{})
+	if !ok {
+		err = fmt.Errorf("cannot parse OpenStack metadata, no object found")
+		return "", err
+	}
+	tmp, ok := obj["uuid"]
+	if !ok {
+		err = fmt.Errorf("cannot parse OpenStack metadata, no uuid found")
+		return "", err
+	}
+	uuid, ok := tmp.(string)
+	if !ok {
+		err = fmt.Errorf("cannot parse OpenStack metadata, uuid has invalid format")
+		return "", err
+	}
+	return uuid, nil
+}
+
+func readInstanceID() (string, error) {
+	// Try to find instance ID on the local filesystem (created by cloud-init)
+	const instanceIDFile = "/var/lib/cloud/data/instance-id"
+	idBytes, err := ioutil.ReadFile(instanceIDFile)
+	if err == nil {
+		instanceID := string(idBytes)
+		instanceID = strings.TrimSpace(instanceID)
+		glog.V(3).Infof("Got instance id from %s: %s", instanceIDFile, instanceID)
+		if instanceID != "" {
+			return instanceID, nil
+		}
+		// Fall through with empty instanceID and try metadata server.
+	}
+	glog.V(5).Infof("Cannot read %s: '%v', trying metadata server", instanceIDFile, err)
+
+	// Try to get JSON from metdata server.
+	const metadataUrl = "http://169.254.169.254/openstack/2012-08-10//meta_data.json"
+	resp, err := http.Get(metadataUrl)
+	if err != nil {
+		glog.V(3).Infof("Cannot read %s: %v", metadataUrl, err)
+		return "", err
+	}
+
+	if resp.StatusCode != 200 {
+		err = fmt.Errorf("got unexpected status code when reading metadata from %s: %s", metadataUrl, resp.Status)
+		glog.V(3).Infof("%v", err)
+		return "", err
+	}
+
+	// OpenStack 'Kilo' sends 'text/html; charset=UTF-8', be prepared for text/plain and no
+	// conent-type at all.
+	cType := resp.Header.Get("Content-Type")
+	if cType != "" && !strings.HasPrefix(cType, "text/plain") && !strings.HasPrefix(cType, "text/html") {
```

Comment:
drop the content type check... I only wanted it if it was sane... the returned content type has nothing to do with the actual content (for example, this would break if the server ever started returning "application/json")